packages/pigeon/lib/java_generator.dart

```diff
@@ -489,21 +561,24 @@ void generateJava(JavaOptions options, Root root, StringSink sink) {
           }
           indent.writeln('return toMapResult;');
         });
-        indent.write('static ${klass.name} fromMap(Map<String, Object> map) ');
+        indent.write(
+            'static @NonNull ${klass.name} fromMap(@NonNull Map<String, Object> map) ');
         indent.scoped('{', '}', () {
-          indent.writeln('${klass.name} fromMapResult = new ${klass.name}();');
+          const String result = 'pigeonResult';
+          indent.writeln('${klass.name} $result = new ${klass.name}();');
           for (final NamedType field in klass.fields) {
             final String fieldVariable = field.name;
+            final String setter = _makeSetter(field);
             indent.writeln('Object $fieldVariable = map.get("${field.name}");');
             if (rootEnumNameSet.contains(field.type.baseName)) {
               indent.writeln(
-                  'fromMapResult.${field.name} = $fieldVariable == null ? null : ${field.type.baseName}.values()[(int)$fieldVariable];');
+                  '$result.$setter($fieldVariable == null ? null : ${field.type.baseName}.values()[(int)$fieldVariable]);');
             } else {
               indent.writeln(
-                  'fromMapResult.${field.name} = ${_castObject(field, root.classes, root.enums, fieldVariable)};');
+                  '$result.$setter(${_castObject(field, root.classes, root.enums, fieldVariable)});');
             }
           }
-          indent.writeln('return fromMapResult;');
+          indent.writeln('return $result;');
         });
```

Comment:
As does some section of this (maybe the loop body, or maybe the bit before plus the loop; it's hard to tell—thus the comment :)—what the coherent sections are to see where best to cut it.